core/src/avm2/object/textformat_object.rs

```diff
@@ -0,0 +1,102 @@
+//! Object representation for TextFormat
+
+use crate::avm2::activation::Activation;
+use crate::avm2::object::script_object::ScriptObjectData;
+use crate::avm2::object::{ClassObject, Object, ObjectPtr, TObject};
+use crate::avm2::value::Value;
+use crate::avm2::Error;
+use crate::html::TextFormat;
+use gc_arena::{Collect, GcCell, MutationContext};
+use std::cell::{Ref, RefMut};
+
+/// A class instance allocator that allocates TextFormat objects.
+pub fn textformat_allocator<'gc>(
+    class: ClassObject<'gc>,
+    proto: Object<'gc>,
+    activation: &mut Activation<'_, 'gc, '_>,
+) -> Result<Object<'gc>, Error> {
+    let base = ScriptObjectData::base_new(Some(proto), Some(class));
+
+    Ok(TextFormatObject(GcCell::allocate(
+        activation.context.gc_context,
+        TextFormatObjectData {
+            base,
+            text_format: Default::default(),
+        },
+    ))
+    .into())
+}
+
+#[derive(Clone, Collect, Debug, Copy)]
+#[collect(no_drop)]
+pub struct TextFormatObject<'gc>(GcCell<'gc, TextFormatObjectData<'gc>>);
+
+#[derive(Clone, Collect, Debug)]
+#[collect(no_drop)]
+pub struct TextFormatObjectData<'gc> {
+    /// Base script object
+    base: ScriptObjectData<'gc>,
+
+    text_format: TextFormat,
+}
+
+impl<'gc> TextFormatObject<'gc> {
+    pub fn from_text_format(
+        activation: &mut Activation<'_, 'gc, '_>,
+        text_format: TextFormat,
+    ) -> Result<Object<'gc>, Error> {
+        let class = activation.avm2().classes().textformat;
+        let proto = activation.avm2().prototypes().textformat;
+        let base = ScriptObjectData::base_new(Some(proto), Some(class));
+
+        let mut this: Object<'gc> = Self(GcCell::allocate(
+            activation.context.gc_context,
+            TextFormatObjectData { base, text_format },
+        ))
+        .into();
+        this.install_instance_traits(activation, class)?;
+
```

Comment:
Note: I've seen many functions (e.g. `RegExpObject::from_regexp`) do the following:
```rs
class.call_native_init(Some(this), &[], activation)?;
```
But I think this is wrong in this case, as it applies the `TextFormat` constructor on the newly-created `this` with no parameters, which deletes its fields.
Maybe I'm missing something, but the existing `call_native_init` calls seem redundant, and even wrong, to me.

Reply:
The pattern I adopted with `StageObject` and a few other object types was that `this` holds nullable data and the initializer only changes anything if that data is `None`. `RegExpObject` does not have a nullable `regexp` field, so this pattern doesn't work. Indeed, you are correct that calling `from_regexp` will not actually use the given `RegExp` correctly. However, this function is apparently unused, which is why this doesn't get flagged by any tests.

Generally speaking associated constructors (such as `from_regexp`) do need to call the class initializer, in one of two cases:

1. The class the object is being allocated as is user-specified.
2. One or more of the object's superclasses (most notably, `EventDispatcher`) does something not accounted for by the associated constructor.

AFAIK neither condition applies for `TextFormatObject` or `RegExpObject`. I added calls to `call_native_init` everywhere for the sake of consistency (e.g. every object allocation also runs class initializers); however it probably would be permissible to drop this for specific classes as long as we document that initializers are being skipped.